transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/jobscheduling/Uploader.java

```diff
@@ -97,50 +97,56 @@ public void upload(TransportContext transportContext, int attemptNumber, Runnabl
 
   void logAndUpdateState(TransportContext transportContext, int attemptNumber) {
     TransportBackend backend = backendRegistry.get(transportContext.getBackendName());
+    long maxNextRequestWaitMillis = 0;
 
-    Iterable<PersistedEvent> persistedEvents =
-        guard.runCriticalSection(() -> eventStore.loadBatch(transportContext));
+    while (guard.runCriticalSection(() -> eventStore.hasPendingEventsFor(transportContext))) {
+      Iterable<PersistedEvent> persistedEvents =
+          guard.runCriticalSection(() -> eventStore.loadBatch(transportContext));
 
-    // Do not make a call to the backend if the list is empty.
-    if (!persistedEvents.iterator().hasNext()) {
-      return;
-    }
+      // Do not make a call to the backend if the list is empty.
+      if (!persistedEvents.iterator().hasNext()) {
+        return;
+      }
 
-    BackendResponse response;
-    if (backend == null) {
-      Logging.d(
-          LOG_TAG, "Unknown backend for %s, deleting event batch for it...", transportContext);
-      response = BackendResponse.fatalError();
-    } else {
-      List<EventInternal> eventInternals = new ArrayList<>();
+      BackendResponse response;
+      if (backend == null) {
+        Logging.d(
+            LOG_TAG, "Unknown backend for %s, deleting event batch for it...", transportContext);
+        response = BackendResponse.fatalError();
+      } else {
+        List<EventInternal> eventInternals = new ArrayList<>();
 
-      for (PersistedEvent persistedEvent : persistedEvents) {
-        eventInternals.add(persistedEvent.getEvent());
+        for (PersistedEvent persistedEvent : persistedEvents) {
+          eventInternals.add(persistedEvent.getEvent());
+        }
+        response =
+            backend.send(
+                BackendRequest.builder()
+                    .setEvents(eventInternals)
+                    .setExtras(transportContext.getExtras())
+                    .build());
+      }
+      if (response.getStatus() == BackendResponse.Status.TRANSIENT_ERROR) {
+        guard.runCriticalSection(
+            () -> {
+              eventStore.recordFailure(persistedEvents);
+              return null;
+            });
+        eventStore.recordNextCallTime(transportContext, clock.getTime() + maxNextRequestWaitMillis);
```

Comment:
Any reason these `eventStore` interactions are not part of the same critical section?

Reply:
Nope, that's a mistake

Reply:
Ah sorry, I force pushed and wiped this out but the only change was wrapping this in the critical section